addons/project/tests/test_project_milestone.py

```diff
@@ -97,3 +142,63 @@ def test_duplicate_project_duplicates_milestones_on_tasks(self):
             self.assertEqual(orig_task.milestone_id.name, copied_task.milestone_id.name,
                              "the copied milestone should be a copy if the original ")
+
+    def test_set_milestone_parent_task(self):
+        """ When a milestone is set on a parent task, it is set as well on its child tasks if they have no milestone set yet and
+            if they belong to the same project (or they have no project set).
+
+            Test Case:
+            =========
+            1) Set a milestone on the task (or not)
+            2) Change the milestone of its parent
+            3) Check the result
+        """
+        # A. Child task with no milestone set and belonging to the same project
+        task_2, task_3 = self.env['project.task'].with_context({'mail_create_nolog': True}).create([{
+            'name': 'Child MilestoneTask',
+            'user_ids': self.user_projectmanager,
+            'project_id': self.project_pigs.id,
+            'parent_id': self.task_1.id,
+        }, {
+            'name': 'Grand-child MilestoneTask',
+            'user_ids': self.user_projectmanager,
+            'project_id': self.project_pigs.id,
+        }])
+        self.assertFalse(self.task_1.milestone_id)
+        self.assertFalse(task_2.milestone_id)
+
+        self.task_1.milestone_id = self.milestone_pigs
+        self.assertEqual(task_2.milestone_id, self.milestone_pigs,
+                         "The milestone of the parent task should be set to its subtasks if they belong to the same project (or the subtask has not project set) and the subtask has no milestone already set.")
+
+        # B. Child task with a milestone already set
+        self.extra_milestone_pigs = self.env['project.milestone'].with_context({'mail_create_nolog': True}).create({
+            'name': 'Test Extra Milestone',
+            'project_id': self.project_pigs.id,
+        })
+        self.task_1.milestone_id = False
+        task_2.milestone_id = self.extra_milestone_pigs
+        self.assertFalse(self.task_1.milestone_id)
+        self.assertEqual(task_2.milestone_id, self.extra_milestone_pigs)
+
+        self.task_1.milestone_id = self.milestone_pigs
+        self.assertEqual(task_2.milestone_id, self.extra_milestone_pigs, "The milestone of the child task should not be modified has it has already one set.")
+
+        # C. Child task with no milestone set but belonging to another project
+        self.task_1.project_id = self.project_goats
+        task_2.milestone_id = False
+        self.assertFalse(self.task_1.milestone_id)
+        self.assertFalse(task_2.milestone_id)
+
+        self.task_1.milestone_id = self.milestone_goats
+        self.assertFalse(task_2.milestone_id, "The milestone of the parent task should not be set to its child task has they belong to different projects.")
+
+        # D. Recursion test (grand-parent task's milestone set to grand-child task)
+        task_3.parent_id = task_2
```

Comment:
why not do that when creating ``task_3`` ? Saves one write operation :shrug:

Reply:
> why not do that when creating `task_3` ? Saves one write operation shrug

The thing is task_2 does not exist yet when I create task_3 (because they are created by the same call to `create`). So either I create them with two operations, either I create them in one operation and write the `parent_id` in a second one. I have honestly no preference and I'm not sure if one way is more optimal or not :man_shrugging:

Reply:
Yeah, you're right here as well. Leave it as is :sunglasses: